File: frontend/lib/types/task.ts
import type { Tag } from "./tag";
import type { Priority } from "./priority";

export type TaskStatus = "pending" | "completed";

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority_id: string | null;
  priority: Priority | null;
  tags: Tag[];
  due_date: string | null;
  recurrence_rule: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority_id?: string;
  tag_ids?: string[];
  due_date?: string;
  recurrence_rule?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string | null;
  status?: TaskStatus;
  priority_id?: string | null;
  tag_ids?: string[];
  due_date?: string | null;
  recurrence_rule?: string | null;
}

export interface TaskFilters {
  status?: TaskStatus | "all";
  priority_id?: string;
  tag_id?: string;
  due_before?: string;
  due_after?: string;
  search?: string;
}

export type SortField = "created_at" | "due_date" | "title" | "priority";
export type SortOrder = "asc" | "desc";

export interface TaskSort {
  sort_by: SortField;
  sort_order: SortOrder;
}

export interface TaskPagination {
  page: number;
  page_size: number;
}

export interface TaskQueryParams extends TaskFilters, TaskSort, TaskPagination {}

export type RecurrenceFrequency = "daily" | "weekly" | "monthly";

export interface RecurrenceRule {
  frequency: RecurrenceFrequency;
  interval?: number;
  end_date?: string;
}

export const RECURRENCE_LABELS: Record<RecurrenceFrequency, string> = {
  daily: "Daily",
  weekly: "Weekly",
  monthly: "Monthly",
};

File: backend/app/schemas/task.py
"""Task Pydantic schemas for API request/response."""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field

from app.models.task import TaskStatus


class TaskCreate(BaseModel):
    """Schema for creating a new task."""
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=5000)
    priority_id: Optional[str] = None
    due_date: Optional[datetime] = None
    recurrence_rule: Optional[str] = Field(
        None,
        pattern="^(daily|weekly|monthly|yearly)$"
    )
    tag_ids: Optional[List[str]] = None


class TaskUpdate(BaseModel):
    """Schema for updating an existing task."""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=5000)
    priority_id: Optional[str] = None
    due_date: Optional[datetime] = None
    recurrence_rule: Optional[str] = Field(
        None,
        pattern="^(daily|weekly|monthly|yearly)$"
    )
    tag_ids: Optional[List[str]] = None


class PriorityInTask(BaseModel):
    """Embedded priority in task response."""
    id: str
    level: str

    class Config:
        from_attributes = True


class TagInTask(BaseModel):
    """Embedded tag in task response."""
    id: str
    name: str

    class Config:
        from_attributes = True


class TaskResponse(BaseModel):
    """Task response schema."""
    id: str
    user_id: str
    title: str
    description: Optional[str] = None
    status: TaskStatus
    priority: Optional[PriorityInTask] = None
    tags: List[TagInTask] = []
    due_date: Optional[datetime] = None
    recurrence_rule: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskListResponse(BaseModel):
    """Paginated task list response."""
    tasks: List[TaskResponse]
    total: int
    page: int
    page_size: int
